File: 2024/day2/2.py
from common.utils import problem_harness, timeit, read_input


def is_safe(level: list) -> bool:
  diff = [level[i] - level[i+1] for i in range(len(level)-1)]
  max_diff = max([abs(level[i] - level[i+1]) for i in range(len(level)-1)])
  if not 1 <= max_diff <= 3:
    return False
  return all(d > 0 for d in diff) or all(d < 0 for d in diff)


def is_safe_with_remove(level: list) -> bool:
  if is_safe(level):
    return True
  return any(is_safe(level[:i] + level[i+1:]) for i in range(len(level)))


@timeit
def part1(filename: str) -> int:
  return sum(
    [is_safe([int(x) for x in line]) for line in read_input(filename, sep=" ")]
  )


@timeit
def part2(filename: str) -> int:
  return sum(
    [is_safe_with_remove(
      [int(x) for x in l]) for l in read_input(filename, sep=" ")]
  )


def main():
  problem_harness(part1, part2)


if __name__ == '__main__':
  main()

File: 2024/day2/2.go
package main

import (
	"aoc/utils"
	"math"
	"strconv"
	"strings"
)

func isGoodLevel(level []int) bool {
	for i := range level {
		if i == 0 {
			continue
		}
		d := level[i] - level[i-1]
		if math.Abs(float64(d)) > 3 || math.Abs(float64(d)) == 0 {
			return false
		}
	}
	return isIncreasing(level) || isDecreasing(level)
}

func isIncreasing(level []int) bool {
	for i := range level {
		if i == 0 {
			continue
		}
		if level[i] < level[i-1] {
			return false
		}
	}
	return true
}

func isDecreasing(level []int) bool {
	for i := range level {
		if i == 0 {
			continue
		}
		if level[i] > level[i-1] {
			return false
		}
	}
	return true
}

func isGoodLevelWithRemoval(level []int) bool {
	// this is on^2 but it's small enough to not matter
	if isGoodLevel(level) {
		return true
	}
	for i := range level {
		// remove element and check if it is good without it
		levelCopy := make([]int, len(level))
		copy(levelCopy, level)
		levelCopy = append(levelCopy[:i], levelCopy[i+1:]...)
		if isGoodLevel(levelCopy) {
			return true
		}
	}
	return false
}

func toInt(s string) int { 
	i, err := strconv.Atoi(s)
	if err != nil {
		panic(err)
	}
	return i
}

func part1(data string) int {
	// I know this isn't very go'idomatic but I like it so idc
	return utils.Reduce(
		strings.Split(data, "\n"), func(acc int, line string) int {
			level := utils.Map(strings.Split(line, " "), toInt)
			if isGoodLevel(level) {
				return acc + 1;
			}
			return acc
		}, 0)
}

func part2(data string) int {
	return utils.Reduce(
		strings.Split(data, "\n"), func(acc int, line string) int {
			level := utils.Map(strings.Split(line, " "), toInt)
			if isGoodLevelWithRemoval(level) {
				return acc + 1;
			}
			return acc
		}, 0)
}

func main() {
	utils.ProblemSetUp(part1, part2)
}
